bin/main.dart

```diff
@@ -58,58 +87,142 @@ _translateLocalizedFiles(List<String> langCodes, String dirPath) async {
-            if (!newLangStrMap.containsKey(targetLang)) {
-              newLangStrMap[targetLang] = Map();
-            }
-            updated = true;
-            newLangStrMap[targetLang][sourceKey] = (await gtr
-                    .translate(sourceString, from: sourceLang, to: targetLang))
-                .text;
+          if (sourceStringMap[sourceKey].isNotEmpty &&
+              !targetStringMap.containsKey(sourceKey)) {
+            var tuple = Tuple2(targetLang, sourceLang);
+            if (toTranslateMap[tuple] == null) toTranslateMap[tuple] = List();
+            toTranslateMap[tuple].add(sourceKey);
           }
         }
       });
     });
   });
-
-  if (!updated) {
-    stdout.writeln('No strings were translated. Exiting...');
+  if (toTranslateMap.isEmpty) {
+    stdout.writeln('No string needs to be translated. Exiting...');
     exit(0);
   }
 
-  oldLangStrMap.forEach((lang, langMap) {
-    if (newLangStrMap.containsKey(lang)) {
-      oldLangStrMap[lang].addAll(newLangStrMap[lang]);
+  // to check performance
+  String date = DateTime.now().toString();
+  stdout.writeln('Translation starts - $date');
+
+  // different functions and arguments for different providers
+  if (provider.compareTo('GoogleTest') == 0) {
+    await _translateGoogleTest(langStringMap, toTranslateMap);
+  } else if (provider.compareTo('Google') == 0) {
+    await _translateGoogle(
+        langStringMap, toTranslateMap, projectKey, numStringsAtOnce);
+  }
+
+  // to check performance
+  date = DateTime.now().toString();
+  stdout.writeln('Translation ends - $date');
+  _updateContent(langStringMap, dirPath);
+}
+
+// see https://github.com/gabrielpacheco23, thanks to Gabriel Pacheco
+_translateGoogleTest(Map<String, Map<String, String>> langStringMap,
+    Map<Tuple2<String, String>, List<String>> toTranslateMap) async {
+  final gtr = GoogleTranslator();
+
+  await Future.forEach(toTranslateMap.entries, (toTranslate) async {
+    final targetLang = toTranslate.key.item1;
+    final sourceLang = toTranslate.key.item2;
+    final keyList = toTranslate.value;
+    await Future.forEach(keyList, (key) async {
+      final sourceString = langStringMap[sourceLang][key];
+      langStringMap[targetLang][key] =
+          (await gtr.translate(sourceString, from: sourceLang, to: targetLang))
+              .text;
+    });
+  });
+}
+
+// Google Translate
+// langStringMap is a map of a language to <key, string>
+// toTranslateMap is a map of <targetLang, sourceLang> to List<key>> - strings to translate
+_translateGoogle(
+    Map<String, Map<String, String>> langStringMap,
+    Map<Tuple2<String, String>, List<String>> toTranslateMap,
+    String projectKey,
+    int numStringsAtOnce) async {
+  await Future.forEach(toTranslateMap.entries, (toTranslate) async {
+    final targetLang = toTranslate.key.item1;
+    final sourceLang = toTranslate.key.item2;
+    final keyList = toTranslate.value;
+    List<String> stringInOutList = List();
+    int num = 0;
```

Comment:
Also a Dart feature. Instead of specifying the type "int" explicitly, while you see the "0" as value is an integer already you can do like this:

 ```suggestion
    var num = 0;
```

Dart is clever and will understand that it is int. (Again during Dart linting it can show a warning)
You may try to integrate this package: https://pub.dev/packages/pedantic